File: goldenflow/transforms/dates.py
from __future__ import annotations

from datetime import date, timedelta

from dateutil import parser as dateutil_parser
import polars as pl

from goldenflow.transforms import register_transform


def _parse_date(val: str | None) -> date | None:
    if not val:
        return None
    try:
        return dateutil_parser.parse(val).date()
    except (ValueError, OverflowError):
        return None


@register_transform(
    name="date_iso8601", input_types=["date"], auto_apply=True, priority=50, mode="series"
)
def date_iso8601(series: pl.Series) -> pl.Series:
    def _fmt(val: str | None) -> str | None:
        if val is None:
            return None
        d = _parse_date(val)
        return d.isoformat() if d else val

    return series.map_elements(_fmt, return_dtype=pl.Utf8)


@register_transform(
    name="date_us", input_types=["date"], auto_apply=False, priority=50, mode="series"
)
def date_us(series: pl.Series) -> pl.Series:
    def _fmt(val: str | None) -> str | None:
        if val is None:
            return None
        d = _parse_date(val)
        return d.strftime("%m/%d/%Y") if d else val

    return series.map_elements(_fmt, return_dtype=pl.Utf8)


@register_transform(
    name="date_eu", input_types=["date"], auto_apply=False, priority=50, mode="series"
)
def date_eu(series: pl.Series) -> pl.Series:
    def _fmt(val: str | None) -> str | None:
        if val is None:
            return None
        d = _parse_date(val)
        return d.strftime("%d/%m/%Y") if d else val

    return series.map_elements(_fmt, return_dtype=pl.Utf8)


@register_transform(
    name="date_parse", input_types=["date"], auto_apply=False, priority=55, mode="series"
)
def date_parse(series: pl.Series) -> pl.Series:
    """Auto-detect format and normalize to ISO 8601."""
    return date_iso8601(series)


@register_transform(
    name="age_from_dob", input_types=["date"], auto_apply=False, priority=40, mode="series"
)
def age_from_dob(series: pl.Series, reference_date: str | None = None) -> pl.Series:
    ref = (
        dateutil_parser.parse(reference_date).date()
        if reference_date
        else date.today()
    )

    def _age(val: str | None) -> int | None:
        if val is None:
            return None
        d = _parse_date(val)
        if d is None:
            return None
        age = ref.year - d.year - ((ref.month, ref.day) < (d.month, d.day))
        return age

    return series.map_elements(_age, return_dtype=pl.Int64)


@register_transform(
    name="datetime_iso8601",
    input_types=["date"],
    auto_apply=False,
    priority=50,
    mode="series",
)
def datetime_iso8601(series: pl.Series) -> pl.Series:
    """Parse to ISO 8601 datetime (with time component)."""

    def _fmt(val: str | None) -> str | None:
        if val is None:
            return None
        try:
            dt = dateutil_parser.parse(val)
            return dt.strftime("%Y-%m-%dT%H:%M:%S")
        except (ValueError, OverflowError):
            return val

    return series.map_elements(_fmt, return_dtype=pl.Utf8)


@register_transform(
    name="extract_year",
    input_types=["date"],
    auto_apply=False,
    priority=35,
    mode="series",
)
def extract_year(series: pl.Series) -> pl.Series:
    """Extract the year as an integer."""

    def _year(val: str | None) -> int | None:
        if val is None:
            return None
        d = _parse_date(val)
        return d.year if d else None

    return series.map_elements(_year, return_dtype=pl.Int64)


@register_transform(
    name="extract_month",
    input_types=["date"],
    auto_apply=False,
    priority=35,
    mode="series",
)
def extract_month(series: pl.Series) -> pl.Series:
    """Extract the month as an integer (1-12)."""

    def _month(val: str | None) -> int | None:
        if val is None:
            return None
        d = _parse_date(val)
        return d.month if d else None

    return series.map_elements(_month, return_dtype=pl.Int64)


@register_transform(
    name="date_shift",
    input_types=["date"],
    auto_apply=False,
    priority=30,
    mode="series",
)
def date_shift(series: pl.Series, days: int = 0) -> pl.Series:
    """Shift dates by a number of days (positive = forward, negative = backward)."""

    def _shift(val: str | None) -> str | None:
        if val is None:
            return None
        d = _parse_date(val)
        if d is None:
            return val
        shifted = d + timedelta(days=days)
        return shifted.isoformat()

    return series.map_elements(_shift, return_dtype=pl.Utf8)


_DAY_NAMES = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]


@register_transform(
    name="extract_day",
    input_types=["date"],
    auto_apply=False,
    priority=35,
    mode="series",
)
def extract_day(series: pl.Series) -> pl.Series:
    """Extract the day of month as an integer (1-31)."""

    def _day(val: str | None) -> int | None:
        if val is None:
            return None
        d = _parse_date(val)
        return d.day if d else None

    return series.map_elements(_day, return_dtype=pl.Int64)


@register_transform(
    name="extract_quarter",
    input_types=["date"],
    auto_apply=False,
    priority=35,
    mode="series",
)
def extract_quarter(series: pl.Series) -> pl.Series:
    """Extract the quarter (1-4) from a date."""

    def _quarter(val: str | None) -> int | None:
        if val is None:
            return None
        d = _parse_date(val)
        if d is None:
            return None
        return (d.month - 1) // 3 + 1

    return series.map_elements(_quarter, return_dtype=pl.Int64)


@register_transform(
    name="extract_day_of_week",
    input_types=["date"],
    auto_apply=False,
    priority=35,
    mode="series",
)
def extract_day_of_week(series: pl.Series) -> pl.Series:
    """Extract the day of week name (Monday, Tuesday, etc.)."""

    def _dow(val: str | None) -> str | None:
        if val is None:
            return None
        d = _parse_date(val)
        if d is None:
            return None
        return _DAY_NAMES[d.weekday()]

    return series.map_elements(_dow, return_dtype=pl.Utf8)


@register_transform(
    name="date_validate",
    input_types=["date", "string"],
    auto_apply=False,
    priority=60,
    mode="series",
)
def date_validate(series: pl.Series) -> pl.Series:
    """Validate if value is a parseable date. Returns True/False/None."""

    def _validate(val: str | None) -> bool | None:
        if val is None:
            return None
        if not val.strip():
            return False
        return _parse_date(val) is not None

    return series.map_elements(_validate, return_dtype=pl.Boolean)

File: packages/goldenflow-js/src/core/transforms/dates.ts
/**
 * Date transforms — ported from goldenflow/transforms/dates.py
 * Side-effect module: registers 13 date transforms on import.
 *
 * All date operations use UTC to avoid timezone-dependent results
 * (matching Python's timezone-naive dateutil behavior).
 */

import type { ColumnValue } from "../types.js";
import { registerTransform } from "./registry.js";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function _parseDate(val: string): Date | null {
  const trimmed = val.trim();
  if (!trimmed) return null;
  const d = new Date(trimmed);
  if (isNaN(d.getTime())) return null;
  return d;
}

function pad(n: number): string {
  return n < 10 ? `0${n}` : String(n);
}

const DAY_NAMES = [
  "Sunday",
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
];

// ---------------------------------------------------------------------------
// date_iso8601 (series, date, 50, auto_apply) — parse -> YYYY-MM-DD
// ---------------------------------------------------------------------------

function dateIso8601(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const s = String(v);
    const d = _parseDate(s);
    if (!d) return v;
    return `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}`;
  });
}

registerTransform(
  { name: "date_iso8601", inputTypes: ["date"], autoApply: true, priority: 50, mode: "series" },
  dateIso8601,
);

// ---------------------------------------------------------------------------
// date_us (series, date, 50) — parse -> MM/DD/YYYY
// ---------------------------------------------------------------------------

function dateUs(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const s = String(v);
    const d = _parseDate(s);
    if (!d) return v;
    return `${pad(d.getUTCMonth() + 1)}/${pad(d.getUTCDate())}/${d.getUTCFullYear()}`;
  });
}

registerTransform(
  { name: "date_us", inputTypes: ["date"], priority: 50, mode: "series" },
  dateUs,
);

// ---------------------------------------------------------------------------
// date_eu (series, date, 50) — parse -> DD/MM/YYYY
// ---------------------------------------------------------------------------

function dateEu(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const s = String(v);
    const d = _parseDate(s);
    if (!d) return v;
    return `${pad(d.getUTCDate())}/${pad(d.getUTCMonth() + 1)}/${d.getUTCFullYear()}`;
  });
}

registerTransform(
  { name: "date_eu", inputTypes: ["date"], priority: 50, mode: "series" },
  dateEu,
);

// ---------------------------------------------------------------------------
// date_parse (series, date, 55) — same as date_iso8601
// ---------------------------------------------------------------------------

registerTransform(
  { name: "date_parse", inputTypes: ["date"], priority: 55, mode: "series" },
  dateIso8601,
);

// ---------------------------------------------------------------------------
// age_from_dob (series, date, 40, param: reference_date=null)
// ---------------------------------------------------------------------------

function ageFromDob(
  values: readonly ColumnValue[],
  referenceDate: unknown = null,
): ColumnValue[] {
  const ref = referenceDate ? _parseDate(String(referenceDate)) : new Date();
  if (!ref) return values.slice();

  return values.map((v) => {
    if (v === null) return null;
    const dob = _parseDate(String(v));
    if (!dob) return v;

    let age = ref.getUTCFullYear() - dob.getUTCFullYear();
    const monthDiff = ref.getUTCMonth() - dob.getUTCMonth();
    if (monthDiff < 0 || (monthDiff === 0 && ref.getUTCDate() < dob.getUTCDate())) {
      age--;
    }
    return age;
  });
}

registerTransform(
  { name: "age_from_dob", inputTypes: ["date"], priority: 40, mode: "series" },
  ageFromDob,
);

// ---------------------------------------------------------------------------
// datetime_iso8601 (series, date, 50) — parse -> YYYY-MM-DDTHH:MM:SS
// ---------------------------------------------------------------------------

function datetimeIso8601(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const d = _parseDate(String(v));
    if (!d) return v;
    return (
      `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}` +
      `T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}`
    );
  });
}

registerTransform(
  { name: "datetime_iso8601", inputTypes: ["date"], priority: 50, mode: "series" },
  datetimeIso8601,
);

// ---------------------------------------------------------------------------
// extract_year/month/day/quarter/day_of_week (series, date, 35)
// ---------------------------------------------------------------------------

function extractYear(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const d = _parseDate(String(v));
    return d ? d.getUTCFullYear() : v;
  });
}

function extractMonth(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const d = _parseDate(String(v));
    return d ? d.getUTCMonth() + 1 : v;
  });
}

function extractDay(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const d = _parseDate(String(v));
    return d ? d.getUTCDate() : v;
  });
}

function extractQuarter(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const d = _parseDate(String(v));
    if (!d) return v;
    return Math.floor(d.getUTCMonth() / 3) + 1;
  });
}

function extractDayOfWeek(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const d = _parseDate(String(v));
    return d ? DAY_NAMES[d.getUTCDay()]! : v;
  });
}

registerTransform({ name: "extract_year", inputTypes: ["date"], priority: 35, mode: "series" }, extractYear);
registerTransform({ name: "extract_month", inputTypes: ["date"], priority: 35, mode: "series" }, extractMonth);
registerTransform({ name: "extract_day", inputTypes: ["date"], priority: 35, mode: "series" }, extractDay);
registerTransform({ name: "extract_quarter", inputTypes: ["date"], priority: 35, mode: "series" }, extractQuarter);
registerTransform({ name: "extract_day_of_week", inputTypes: ["date"], priority: 35, mode: "series" }, extractDayOfWeek);

// ---------------------------------------------------------------------------
// date_shift (series, date, 30, param: days=0) — add days, return ISO date
// ---------------------------------------------------------------------------

function dateShift(
  values: readonly ColumnValue[],
  days: unknown = 0,
): ColumnValue[] {
  const shift = typeof days === "number" ? days : Number(days) || 0;
  const shiftMs = shift * 86_400_000;

  return values.map((v) => {
    if (v === null) return null;
    const d = _parseDate(String(v));
    if (!d) return v;
    const shifted = new Date(d.getTime() + shiftMs);
    return `${shifted.getUTCFullYear()}-${pad(shifted.getUTCMonth() + 1)}-${pad(shifted.getUTCDate())}`;
  });
}

registerTransform(
  { name: "date_shift", inputTypes: ["date"], priority: 30, mode: "series" },
  dateShift,
);

// ---------------------------------------------------------------------------
// date_validate (series, date|string, 60) — returns boolean
// ---------------------------------------------------------------------------

function dateValidate(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const s = String(v).trim();
    if (!s) return false;
    return _parseDate(s) !== null;
  });
}

registerTransform(
  { name: "date_validate", inputTypes: ["date", "string"], priority: 60, mode: "series" },
  dateValidate,
);
